File: app/src/main/java/com/novel/read/activity/NovelBookTypeListActivity.java
package com.novel.read.activity;

import android.os.Bundle;

import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.mango.mangolib.event.EventManager;
import com.novel.read.R;
import com.novel.read.adapter.BookListAdapter;
import com.novel.read.base.NovelBaseActivity;
import com.novel.read.constants.Constant;
import com.novel.read.event.SearchListEvent;
import com.novel.read.http.AccountManager;
import com.novel.read.model.protocol.SearchResp;
import com.novel.read.widget.RefreshLayout;
import com.squareup.otto.Subscribe;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

import static com.novel.read.constants.Constant.COMMENT_SIZE;

public class NovelBookTypeListActivity extends NovelBaseActivity {

    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.rlv_type_list)
    RecyclerView mRlvTypeList;
    @BindView(R.id.refresh)
    RefreshLayout refreshLayout;
    private List<SearchResp.BookBean> mList;
    private BookListAdapter mAdapter;
    private String mCategoryId;
    private String mTitle;
    private int page = 1;
    private int loadSize;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_book_type_list;
    }

    @Override
    protected void initView() {
        EventManager.Companion.getInstance().registerSubscriber(this);

        mCategoryId = getIntent().getStringExtra(Constant.Bundle.CategoryId);
        mTitle = getIntent().getStringExtra(Constant.Bundle.mTitle);

        mRlvTypeList.setLayoutManager(new LinearLayoutManager(this));
        mList = new ArrayList<>();

        mAdapter = new BookListAdapter(mList, mRlvTypeList);
        mRlvTypeList.setAdapter(mAdapter);

        mAdapter.setOnLoadMoreListener(() -> {
            if (mAdapter.isLoadingMore()) {

            } else {
                if (loadSize >= COMMENT_SIZE) {
                    mAdapter.setLoadingMore(true);
                    mList.add(null);
                    mAdapter.notifyDataSetChanged();
                    page++;
                    getData();
                }
            }
        });
    }


    @Override
    protected void initData() {
        refreshLayout.showLoading();
        refreshLayout.setOnReloadingListener(this::getData);
        getData();
        toolbar.setTitle(mTitle);
        toolbar.setNavigationOnClickListener(view -> finish());

    }

    private void getData() {
        AccountManager.getInstance().getSearchBookList(mCategoryId, "", page);
    }

    @Subscribe
    public void getSearchList(SearchListEvent event) {
        refreshLayout.showFinish();
        if (event.isFail()) {
            refreshLayout.showError();
        } else {
            loadSize = event.getResult().getBook().size();
            if (mAdapter.isLoadingMore()){
                mList.remove(mList.size() - 1);
                mList.addAll(event.getResult().getBook());
                mAdapter.notifyDataSetChanged();
                mAdapter.setLoadingMore(false);
            }else {
                mList.clear();
                mList.addAll(event.getResult().getBook());
                mAdapter.notifyDataSetChanged();
            }
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        EventManager.Companion.getInstance().unregisterSubscriber(this);
    }
}

File: app/src/main/java/com/novel/read/activity/NovelBookTypeListActivity.kt
package com.novel.read.activity

import androidx.recyclerview.widget.LinearLayoutManager
import com.mango.mangolib.event.EventManager
import com.novel.read.R
import com.novel.read.adapter.BookListAdapter
import com.novel.read.base.NovelBaseActivity
import com.novel.read.constants.Constant
import com.novel.read.constants.Constant.COMMENT_SIZE
import com.novel.read.event.SearchListEvent
import com.novel.read.http.AccountManager
import com.novel.read.inter.OnLoadMoreListener
import com.novel.read.model.protocol.SearchResp
import com.squareup.otto.Subscribe
import kotlinx.android.synthetic.main.activity_book_type_list.*
import java.util.*

class NovelBookTypeListActivity(override val layoutId: Int = R.layout.activity_book_type_list) : NovelBaseActivity() {

    private var mList: MutableList<SearchResp.BookBean> = ArrayList()
    private lateinit var mAdapter: BookListAdapter
    private var mCategoryId: String? = null
    private var mTitle: String? = null
    private var page: Int = 1
    private var loadSize: Int = 0

    override fun initView() {
        EventManager.instance.registerSubscriber(this)

        mCategoryId = intent.getStringExtra(Constant.Bundle.CategoryId)
        mTitle = intent.getStringExtra(Constant.Bundle.mTitle)

        rlv_type_list.layoutManager = LinearLayoutManager(this)
        mAdapter = BookListAdapter(mList, rlv_type_list)
        rlv_type_list.adapter = mAdapter

        mAdapter.setOnLoadMoreListener(object : OnLoadMoreListener {
            override fun onLoadMore() {
                if (mAdapter.isLoadingMore) {
                    rlv_type_list.stopScroll()
                } else {
                    if (loadSize >= COMMENT_SIZE) {
                        mAdapter.isLoadingMore = true
                        mAdapter.notifyDataSetChanged()
                        page++
                        getData()
                    }
                }
            }
        })
    }

    override fun initData() {
        refresh.showLoading()
        refresh.setOnReloadingListener { this.getData() }
        getData()
        toolbar.title = mTitle
        toolbar.setNavigationOnClickListener { finish() }

    }

    private fun getData() {
        mCategoryId?.let { AccountManager.getInstance().getSearchBookList(it, "", page) }
    }

    @Subscribe
    fun getSearchList(event: SearchListEvent) {
        refresh.showFinish()
        if (event.isFail) {
            refresh.showError()
        } else {
            loadSize = event.result!!.book.size
            if (mAdapter.isLoadingMore) {
                mAdapter.isLoadingMore = false
                mList.addAll(event.result!!.book)
                mAdapter.notifyDataSetChanged()
            } else {
                mList.clear()
                mList.addAll(event.result!!.book)
                mAdapter.notifyDataSetChanged()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        EventManager.instance.unregisterSubscriber(this)
    }
}
